File: acwing/src/main/java/dataStructure/TreeArray.java
package dataStructure;

import java.util.Arrays;

/**
 * @author: Zekun Fu
 * @date: 2022/8/10 8:03
 * @Description: 树状数组的模板
 */
public class TreeArray {

    private int[] c;
    private int n;

    public TreeArray(int _n) {
        n = _n;
        c = new int[n + 5];
        Arrays.fill(c, 0);
    }

    private int low_bit(int x) {
        return x & -x;
    }
    public void add(int x, int num) {
        for (int i = x; i <= n; i += low_bit(i)) c[i] += num;
    }
    public int sum(int x) {
        int res = 0;
        for (int i = x; i != 0; i -= low_bit(i)) res += c[i];
        return res;
    }
}

File: codes/leetcodeP/dataStructure/treeArray.py
"""
@author:fuzekun
@file:treeArray.py
@time:2022/08/10
@description:树状数组模板
"""
class TreeArray:

    c = []
    n = 0
    def __init__(self, _n):
        global n, c
        n = _n
        c = [0] * (n + 5)

    def low_bit(self, x:int) :
        return x & -x

    def add(self, x:int, num:int):
        global n, c
        i = x
        while i <= n:
            c[i] += num
            i += self.low_bit(i)

    def sum(self, x:int) :
        global n, c
        i, res = x, 0
        while i > 0:
            res += c[i]
            i -= self.low_bit(i)
        return res